lib/Frontend/Frontend.cpp

```diff
@@ -261,233 +262,300 @@ ModuleDecl *CompilerInstance::getMainModule() {
-  const FrontendOptions &options = Invocation.getFrontendOptions();
-  const InputFileKind Kind = Invocation.getInputKind();
-  ModuleDecl *MainModule = getMainModule();
-  Context->LoadedModules[MainModule->getName()] = MainModule;
-
-  auto modImpKind = SourceFile::ImplicitModuleImportKind::Stdlib;
+static void
+addAdditionalInitialImportsTo(SourceFile *SF,
+                              ModuleDecl *objCModuleUnderlyingMixedFramework,
+                              ModuleDecl *importedHeaderModule,
+                              SmallVectorImpl<ModuleDecl *> &importModules) {
+  SharedTimer timer("performSema-addAdditionalInitialImportsTo");
+  if (!objCModuleUnderlyingMixedFramework && !importedHeaderModule &&
+      importModules.empty())
+    return;
 
-  if (Kind == InputFileKind::IFK_SIL) {
-    assert(BufferIDs.size() == 1);
-    assert(MainBufferID != NO_SUCH_BUFFER);
-    // Assume WMO, if a -primary-file option was not provided.
-    createSILModule(!options.PrimaryInput.hasValue());
-    modImpKind = SourceFile::ImplicitModuleImportKind::None;
-  } else if (Invocation.getParseStdlib()) {
-    modImpKind = SourceFile::ImplicitModuleImportKind::Builtin;
+  using ImportPair =
+      std::pair<ModuleDecl::ImportedModule, SourceFile::ImportOptions>;
+  SmallVector<ImportPair, 4> additionalImports;
+
+  if (objCModuleUnderlyingMixedFramework)
+    additionalImports.push_back(
+        {{/*accessPath=*/{}, objCModuleUnderlyingMixedFramework},
+         SourceFile::ImportFlags::Exported});
+  if (importedHeaderModule)
+    additionalImports.push_back({{/*accessPath=*/{}, importedHeaderModule},
+                                 SourceFile::ImportFlags::Exported});
+  if (!importModules.empty()) {
+    for (auto &importModule : importModules) {
+      additionalImports.push_back({{/*accessPath=*/{}, importModule}, {}});
+    }
   }
 
-  switch (modImpKind) {
+  SF->addImports(additionalImports);
+}
+
+// Implicitly import the SwiftOnoneSupport module in non-optimized
+// builds. This allows for use of popular specialized functions
+// from the standard library, which makes the non-optimized builds
+// execute much faster.
+static bool shouldImplicityImportSwiftOnoneSupportModule(
+    SILOptions::SILOptMode optimization,
+    FrontendOptions::ActionType requestedAction) {
+  return ((optimization <= SILOptions::SILOptMode::None &&
+           (requestedAction == FrontendOptions::EmitObject ||
+            requestedAction == FrontendOptions::Immediate ||
+            requestedAction == FrontendOptions::EmitSIL)) ||
+          (optimization == SILOptions::SILOptMode::None &&
+           requestedAction >= FrontendOptions::EmitSILGen));
+}
+
+void CompilerInstance::performSema() {
+  SharedTimer timer("performSema");
+  Context->LoadedModules[MainModule->getName()] = getMainModule();
+
+  const SourceFile::ImplicitModuleImportKind implicitModuleImportKind =
+      createSILModuleIfNecessary(BufferIDs, MainBufferID);
```

Comment:
These two actions have nothing to do with each other; please don't link them.

Reply:
I like the separation you request. It was because the original code did link them that I left them linked, thinking there was some connection I could not see. 
It might clarify the code to add something that mentions why a SIL module is being created only if the input is a SIL module. Or at least it would help me understand. Would you like to give me some words to put in there?

Reply:
Please take a look at how I've redone this and let me know, tx.